source/common/http/header_map_impl.cc

```diff
@@ -372,6 +384,10 @@ void HeaderMapImpl::addCopy(const LowerCaseString& key, uint64_t value) {
 void HeaderMapImpl::addCopy(const LowerCaseString& key, absl::string_view value) {
   // In the case that the header is appended, we will perform a needless copy of the key and value.
   // This is done on purpose to keep the code simple and should be rare.
+  if (!validHeaderString(key.get()) || !validHeaderString(value)) {
```

Comment:
Where are these coming from? `key` is a `LowerCaseString` so should always be valid.
`value` may not, but it is the caller's responsibility to call this API with valid headers.

Reply:
Oh, you're right! I just added both conditions while trying to get the unit test to work since I wasn't sure which one was causing an assert failure (and noticed that it also fixed the issue in header_parser_fuzz_test), but it seems like it's just `value` that's causing the crashes. Fixed this in the latest commit.

Reply:
Doesn't rejecting the invalid string fix the crash? (We shouldn't be calling addReference/addcopy with invalid strings)

Reply:
Decided to roll back changes related to header_parser and split off into a separate PR. Also updated the unit test to reflect the bug in evaluator_fuzz_test (no longer requires the changes made here).